### PR TITLE
chore: add garbage collector for cleaning in creating snapshots with no transactions

### PR DESCRIPTION
**Pull request by Abhinandan-Purkait**

### CODE REVIEW

File: control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs [hunk 1]
```diff
@@ -181,3 +182,30 @@ async fn prune_volume_snapshot_reconciler(
...
+    // Only proceed if the spec is in creating and has no transactions.
+    if !snapshot.as_ref().status().creating()
+        && !snapshot.as_ref().metadata().transactions().is_empty()
+    {
+        return Ok(PollerState::Idle);
+    }
+
+    let Some(snap_rsc) = context.specs().volume_snapshot_rsc(snapshot.uuid()) else {
+        return Ok(PollerState::Idle);
+    };
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs]: why do we need this, can't we just do:
```suggestion
    let snap_rsc = snapshot.deref().clone();
```

File: control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs [hunk 2]
```diff
@@ -181,3 +182,30 @@ async fn prune_volume_snapshot_reconciler(
...
+
+    snapshot
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs:195]: Should we try to get the volume here and delete through the volume if available?

File: control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs (hunk 1 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs]: We need the ResourceMutex

File: control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs (hunk 2 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs:195]: Good point.

File: control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs (hunk 1 above)
Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs]: Then let's add a way to get it, and fix existing usages too, example:

```patch
diff --git a/control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs b/control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs
index 28c3a0c7..4868f649 100644
--- a/control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs
+++ b/control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs
@@ -4,7 +4,7 @@ use crate::{
         resources::{
             operations::{ResourceLifecycleWithLifetime, ResourcePruning},
             operations_helper::{OperationSequenceGuard, SpecOperationsHelper},
-            OperationGuardArc,
+            OperationGuardArc, ResourceMutex,
         },
         task_poller::{PollContext, PollResult, PollTimer, PollerState, TaskPoller},
     },
@@ -90,10 +90,7 @@ async fn deleting_volume_snapshot_reconciler(
         return Ok(PollerState::Idle);
     }
 
-    let Some(snap_rsc) = context.specs().volume_snapshot_rsc(snapshot.uuid()) else {
-        return Ok(PollerState::Idle);
-    };
-
+    let snap_rsc: ResourceMutex<VolumeSnapshot> = snapshot.resource().clone();
     let snap_uuid = snapshot.uuid().clone();
     match snapshot
         .destroy(
diff --git a/control-plane/agents/src/bin/core/controller/resources/mod.rs b/control-plane/agents/src/bin/core/controller/resources/mod.rs
index 3540e960..a1a3b41c 100644
--- a/control-plane/agents/src/bin/core/controller/resources/mod.rs
+++ b/control-plane/agents/src/bin/core/controller/resources/mod.rs
@@ -197,6 +197,11 @@ impl<T: OperationSequencer + Sized, R> OperationGuard<T, R> {
         }
     }
 }
+impl<T: Clone + Debug + AsOperationSequencer, R> OperationGuard<ResourceMutex<T>, R> {
+    pub fn resource(&self) -> &ResourceMutex<T> {
+        &self.inner
+    }
+}
 
 /// Tracing simple string messages with resource specific information
 /// eg, volume.uuid for volumes and replica.uuid for replicas
```

Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/controller/reconciler/snapshot/garbage_collector.rs]: Done.

**Review by tiagolobocastro — APPROVED** [review]

**Comment by Abhinandan-Purkait** [comment]
bors merge

**Comment by Abhinandan-Purkait** [comment]
bors merge

**Comment by Abhinandan-Purkait** [comment]
bors cancel

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors try

File: control-plane/agents/src/bin/core/volume/snapshot_operations.rs [hunk 3]
```diff
@@ -252,10 +253,20 @@ impl ResourcePruning for OperationGuardArc<VolumeSnapshot> {
...
+                VolumeSnapshotSpecStatus::Creating => transactions > 0,
+                VolumeSnapshotSpecStatus::Created(_) => transactions > 1,
+                _ => false,
+            }
+        }
+
+        if !pruning_allowed(
+            self.as_ref().status(),
+            self.as_ref().metadata().transactions().len(),
+        ) || !self.check_nodes_availability(registry).await
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/snapshot_operations.rs]: why not simply assign a variable here? example:

```suggestion
        let transactions = self.as_ref().metadata().transactions().len();
        let pruning_allowed = match self.as_ref().status() {
            VolumeSnapshotSpecStatus::Creating => transactions > 0,
            VolumeSnapshotSpecStatus::Created(_) => transactions > 1,
            _ => false,
        };

        if !pruning_allowed || !self.check_nodes_availability(registry).await
```

**Review by tiagolobocastro — APPROVED** [review]

File: control-plane/agents/src/bin/core/volume/snapshot_operations.rs (hunk 3 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/snapshot_operations.rs]: Yeah, we can.

**Comment by Abhinandan-Purkait** [comment]
bors merge

**Pull request merged**